app/services/frontend_logger.rb

```diff
@@ -0,0 +1,23 @@
+class FrontendLogger
+  attr_reader :analytics, :event_map
+
+  def initialize(analytics:, event_map:)
+    @analytics = analytics
+    @event_map = event_map
+  end
+
+  def track_event(name, attributes)
+    if (analytics_method = event_map[name])
+      if analytics_method.is_a?(Proc)
+        analytics_method.call(analytics, **attributes)
+      else
+        analytics_method.bind_call(
+          analytics,
+          **MethodSignatureHashBuilder.from_hash(attributes, analytics_method),
```

Comment:
i might even combine the `MethodSignatureHashBuilder` class into this one? since they feel like part of the same work? but still overall very straightforward, big fan of this

Reply:
Yeah, good point. Not sure we'll have a lot of use for `MethodSignatureHashBuilder` outside of this, so I can collapse it.

Reply:
Updated in a326e76d9.

Reply:
🚢 🚢 🚢 🚢 🚢 🚢  love this! let's merge